model-optimizer/extensions/middle/InsertSelect.py

```diff
@@ -36,38 +48,63 @@ def run_before(self):
         return [ReplaceSpliceNodePattern]
 
     @staticmethod
-    def pattern():
-        return dict(
-            nodes=[('op', dict(op='Assign'))],
-            edges=[])
+    def calculate_frame_time(graph: Graph):
+        # in Kaldi can be 1 or 2 inputs. Only main input can change delay in network.
+        # Usually ivector input have name 'ivector'.
+        inputs = graph.get_op_nodes(op='Parameter')
+        if len(inputs) == 1:
+            inp_name = inputs[0].name
+        elif len(inputs) == 2:
+            if inputs[0].name == 'ivector':
+                inp_name = inputs[1].name
+            elif inputs[1].name == 'ivector':
+                inp_name = inputs[0].name
+            else:
+                raise Error("There are 2 inputs for Kaldi model but we can't find out which one is ivector. " +
+                            "Use name \'ivector\' for according input")
+        else:
+            raise Error("There are {} inputs for Kaldi model but we expect only 1 or 2".format(len(inputs)))
+
+        # sort nodes to calculate delays
+        nodes = list(bfs_search(graph, [inp_name]))
+        nx.set_node_attributes(G=graph, name='frame_time', values=-1)
```

Comment:
I think it is not a good idea to use nx routine in transformations. Here instead we can traverse nodes using `for` and set needed attribute

Reply:
yes, we can but why we should?

Reply:
nx is mainly used in case of a real need in low-level API. In transformation our high-level graph API is sufficient.
I propose to avoid nx in transformations.

Reply:
It is a fast way to set an attribute for all nodes in the graph. And this approach does not have any known issues. So I don't have strong objection against using this method.